juju/juju
# feat(azure): add support for managed identities

Azure controllers can now be configured to use user managed identities via the instance-role constraint.

Some small refactoring of the resource group set up was needed since the group now needs to exist for the managed identity to be created.

The feature works the similarly to what was done for AWS - we have an "instance-role" constraint attribute which can be used to specify the user managed identity to use. The constraint naming is a little unfortunate - it's somewhat AWS specific - but it's what we have and we don't want to introduce a new constraint attribute.

The user can create their own user managed identity or ask Juju to do it.

1. Juju creates managed identity
Here, the instance-role is set to "auto".
Juju will create a managed identity in the resource group used to host the controller model.
bootstrap:
`juju bootstrap azure --constraints="instance-role=auto" ...`

2. User supplied managed identity
This is more for advanced users who want to manually manage their Azure assets.
Here, the user needs to create their own managed identity with a role assignment conferring the following allowed actions:
```
"Microsoft.Compute/*",
"Microsoft.KeyVault/*",
"Microsoft.Network/*",
"Microsoft.Resources/*",
"Microsoft.Storage/*",
"Microsoft.ManagedIdentity/userAssignedIdentities/*"
```

The assignable scopes need to be compatible with the intended deployment, eg
```
"concat('/subscriptions/', subscription().subscriptionId)",
"resourceGroup().id"
```

The instance role can be located in a different group or even different subscription to that used to host the Juju controller.
The formats supported are:
`managedinstancename`
`groupname/managaedinstancename`
`subscription/groupname/managedinstancename`

bootstrap:
`juju bootstrap azure --constraints="instance-role=mymanagedinstance" ...`

## QA steps

Test various bootstrap scenarios. In each case, the controller credential can be inspected:
```
$ juju show-credential azure default
controller-credentials:
  azure:
    default:
      content:
        auth-type: instance-role
        validity-check: valid
        managed-identity: juju-controller-2e52ba20-068b-41f9-81d1-1ed9e84e52c6
        subscription-id: 2eebf55a-1e02-45d8-a299-02aed8aea00b
      models:
        controller: admin
```

Also deploy a charm with storage, eg postgresql.
Also run enable-ha for each and ensure all 3 controllers come up.

Scenario one - regression check
`juju bootstrap azure`
(here the controller credential will be a service principal)

Scenario two - juju managed resources
`juju bootstrap azure --constraints="instance-role=auto"`

Scenario three - custom resource group
First create the group, then
`juju bootstrap azure --config resource-group-name=mygroup --constraints="instance-role=auto"`

Scenario four - manual managed identity
Create a managed identity in a resource group
`juju bootstrap azure --config resource-group-name=mygroup --constraints="instance-role=mygroup/identityname"`

Scenario five - manual managed identity different group
Create a managed identity in a resource group
Create a different resource group
`juju bootstrap azure --config resource-group-name=mygroup --constraints="instance-role=othergroup/identityname"`

Bash CI test can be run - these cover the auto and custom managed identity case where the identity is created in some random group different to the group used for the controller.
You'll need to have the az client installed and be logged in to an Azure account.

`./main.sh -v -p azure azure_cloud`

## Documentation changes

@tmihoc the PR description covers the core concepts - we'll need to update the Azure provider doc along the same lines as would have been done for the AWS doc when this feature landed

## Links

**Jira card:** [JUJU-6206](https://warthogs.atlassian.net/browse/JUJU-6206)



[JUJU-6206]: https://warthogs.atlassian.net/browse/JUJU-6206?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
